### PR TITLE
Feat: datepicker

### PR DESCRIPTION
**Pull request by iisakkirotko**

### CODE REVIEW

File: solara/lab/components/datepicker.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import solara.lab
+
+
+@solara.component
+def DatePicker(
+    date: solara.Reactive[dt.date],
+    children: Optional[List[solara.Element]] = [],
+    open: Optional[Union[solara.Reactive[bool], bool]] = False,
+    date_format: str = "%Y-%m-%d",
+):
```
Comment (maartenbreddels) [inline solara/lab/components/datepicker.py:14]: I think this should be named an 'InputDate' similar to the InputInt/Float etc. I think the v-date-picker in vue is maybe not a good name. Looking at MUI https://mui.com/x/react-date-pickers/date-picker/ and https://mui.com/x/react-date-pickers/date-calendar/ it seems like MUI also uses the name 'DatePicker' for what you are making here.
Maybe a better set of names is 'DateCalender' or 'Calendar' and 'InputDate' (we can argue about naming it DateInput later).

In any case, lets not yet expose the Calendar component, and keep that for later.

Comment (iisakkirotko) [reply solara/lab/components/datepicker.py]: `InputDate` and `InputDateRange` it is.

File: solara/lab/components/input_date.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            Tuple[Optional[dt.date], Optional[dt.date]], tuple([dt.datetime.strptime(item, "%Y-%m-%d").date() if item is not None else None for item in values])
+        )
+        value.value = date_value
+
+    input = solara.v.TextField(
+        label=label,
+        value=" - ".join(date_strings),
+        on_v_model=set_dates_cast,
+        append_icon="mdi-calendar",
+        readonly=True,
```
Comment (maartenbreddels) [inline solara/lab/components/input_date.py:170]: I think we want to remove this right, given that we respond to changes (on_v_model)?

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Really great to see this coming to Solara!
For testing, I suggest you look input_test.py. There are a lot of edge cases in those that probably translate to this component as well.

**Comment by iisakkirotko** [comment]
@maartenbreddels let me know what you think about the advanced example.

**Comment by iisakkirotko** [comment]
There is still some behaviour going on that I need to iron out:
- The Menu containing the calendar in the inputs does not close when tab or enter are pressed. Something resets this even when `datepicker_is_open` is set to `False` within `use_change`
- When an invalid date is entered by typing it in to the text field, the date resets to the previous valid one, while retaining error colours. This might be desirable behaviour, depending on who you ask.

**Comment by iisakkirotko** [comment]
Good to know that this relies on PRs #325 and #338.

**Comment by iisakkirotko** [comment]
Two things to note:

- Currently `InputDate` and `InputDateRange` don't support the `blur` event, or closing the datepicker on tab/enter
- `InputDateRange` will be tested on the front-end (playwright), since the `TextField` of that component is read only

**Comment by iisakkirotko** [comment]
Now also depends on https://github.com/widgetti/solara/pull/345

File: solara/lab/components/input_date.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    def Page():
+        date = solara.use_reactive(dt.date.today())
+        range_is_open = solara.use_reactive(False)
+
+        with solara.Column():
+            lab.InputDate(
+                date,
+                open_value=range_is_open,
+            )
+            solara.Text(str(date.value))
```
Comment (maartenbreddels) [inline solara/lab/components/input_date.py:68]:
```suggestion
        lab.InputDate(date)
        solara.Text(str(date.value))
```

**Review by maartenbreddels — CHANGES_REQUESTED** [review]

File: solara/lab/components/input_date.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+                    solara.Button(
+                        label="Book",
+                        color="primary",
+                        on_click=book,
+                    )
+
+    ```
+    """
+    value_reactive = solara.use_reactive(value)
+    del value
```
Comment (maartenbreddels) [inline solara/lab/components/input_date.py:249]:
```suggestion
    value_reactive = solara.use_reactive(value, on_value)  # type: ignore
    del value, on_value
```

File: tests/integration/input_date_test.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import datetime as dt
+from unittest.mock import MagicMock
+
+# import ipyvuetify as vw
```
Comment (maartenbreddels) [inline tests/integration/input_date_test.py:4]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    assert on_v_model.call_count == 1
+    # input.widget.fire_event("blur")
+    assert on_value.call_args[0][0] == tuple([dt.date(2023, 10, 17), None])
+    assert text_field.widget.error
+    assert text_field.widget.label == "label (Please select two dates)"
+
+    input.widget.v_model = ["2023-10-17", tomorrow.strftime("%Y-%m-%d")]
+    assert on_value.call_count == 3
+    assert on_v_model.call_count == 2
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:156]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    text_field = rc.find(vw.TextField)
+    input = rc.find(vw.DatePicker)
+    text_field.widget.observe(on_v_model, "v_model")
+
+    assert input.widget.v_model == [today.strftime("%Y-%m-%d"), tomorrow.strftime("%Y-%m-%d")]
+    input.widget.v_model = [tomorrow.strftime("%Y-%m-%d"), today.strftime("%Y-%m-%d")]
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    on_v_model.reset_mock()
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:141]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    assert on_value.call_args is None
+    assert input.widget.error
+    assert input.widget.v_model == "2023/10"
+    assert on_v_model.call_count == 1
+
+    input.widget.v_model = "2023/10/17"
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 2
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:119]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    el = Test()
+    box, rc = solara.render(el, handle_error=False)
+    input = rc.find(vw.TextField)
+    input.widget.observe(on_v_model, "v_model")
+
+    input.widget.v_model = "2023/10"
+    assert on_value.call_count == 0
+    assert on_v_model.call_count == 1
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:110]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    assert on_value.call_count == 2
+    assert on_v_model.call_count == 2
+    # input.widget.fire_event("blur")
+    assert on_value.call_args[0][0] is None
+    assert not input.widget.error
+
+    input.widget.v_model = "2023/10/17"
+    assert on_value.call_count == 3
+    assert on_v_model.call_count == 3
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:82]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    input.widget.v_model = tomorrow.strftime("%Y/%m/%d")
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    # input.widget.fire_event("blur")
+    assert on_value.call_args[0][0] == tomorrow
+
+    input.widget.v_model = ""
+    assert on_value.call_count == 2
+    assert on_v_model.call_count == 2
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:75]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    on_v_model = MagicMock()
+    el = InputDate(value=today, label="label", on_value=on_value, optional=True)
+    box, rc = solara.render(el, handle_error=False)
+    input = rc.find(vw.TextField)
+    input.widget.observe(on_v_model, "v_model")
+
+    input.widget.v_model = tomorrow.strftime("%Y/%m/%d")
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:69]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+    assert on_value.call_args[0][0] == tomorrow
+    assert input.widget.error
+    assert input.widget.label == "label (Date cannot be empty)"
+
+    input.widget.v_model = "2023/10/17"
+    assert on_value.call_count == 2
+    assert on_v_model.call_count == 2
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:52]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    on_v_model.reset_mock()
+    # input.widget.fire_event("blur")
+    assert on_value.call_args[0][0] == tomorrow
+
+    input.widget.v_model = ""
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:44]:
```suggestion
```

File: tests/unit/input_date_test.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+    el = InputDate(value=today, label="label", on_value=on_value)
+    box, rc = solara.render(el, handle_error=False)
+    input = rc.find(vw.TextField)
+    input.widget.observe(on_v_model, "v_model")
+
+    input.widget.v_model = tomorrow.strftime("%Y/%m/%d")
+    assert on_value.call_count == 1
+    assert on_v_model.call_count == 1
+    on_v_model.reset_mock()
+    # input.widget.fire_event("blur")
```
Comment (maartenbreddels) [inline tests/unit/input_date_test.py:38]:
```suggestion
```

File: solara/lab/components/input_date.py [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+            value_reactive.value = date_value
+
+    def standard_strfy(date: Optional[dt.date]):
+        if date is None:
+            return None
+        else:
+            return date.strftime("%Y-%m-%d")
+
+    value_reactive = solara.use_reactive(value)
+    del value
```
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/lab/components/input_date.py:112]: would be best to move these to the beginning of the code, such that the del's are executed first, and we don't accidentally use the non-reactive value.

**Comment by maartenbreddels** [comment]
🥳

**Comment by iisakkirotko** [comment]
[nice](https://app.graphite.dev/macros/widgetti)

**Pull request merged**